### PR TITLE
BaseProfiler save() for json

### PR DESCRIPTION
**Pull request by micdavis**
* Implemented `_json_save_helper()`
* implemented `save()` to support json
* wrote initial tests

### CODE REVIEW

File: dataprofiler/profilers/profile_builder.py [hunk 1]
```diff
@@ -1531,21 +1546,24 @@ def save(self, filepath: str = None) -> None:
...
-            "_samples_per_update": self._samples_per_update,
-            "_min_true_samples": self._min_true_samples,
-            "_empty_line_count": self._empty_line_count,
-            "memory_size": self.memory_size,
-            "options": self.options,
-            "_profile": self.profile,
-            "times": self.times,
-        }
-        self._save_helper(filepath, data_dict)
+        if filepath is None or filepath.split(".")[-1] != "json":
```
Comment (micdavis) [inline dataprofiler/profilers/profile_builder.py:1549]: basically if the filepath doesn't end in "json" just default to pkl

File: dataprofiler/tests/profilers/test_profile_builder.py [hunk 2]
```diff
@@ -3834,6 +3898,19 @@ def test_save_and_load_no_labeler(self):
         save_profile.update_profile(pd.DataFrame(["test", "test2"]))
         load_profile.update_profile(pd.DataFrame(["test", "test2"]))
 
+    def test_save_json_file(self):
```
Comment (micdavis) [inline dataprofiler/tests/profilers/test_profile_builder.py:3901]: testing `NotImplementedError()`

File: dataprofiler/profilers/profile_builder.py (hunk 1 above)
Comment (micdavis) [reply dataprofiler/profilers/profile_builder.py:1549]: or if the filepath doesn't exist

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -108,7 +108,7 @@ repos:
         additional_dependencies: ['h5py', 'wheel', 'future', 'numpy', 'pandas',
         'python-dateutil', 'pytz', 'pyarrow', 'chardet', 'fastavro',
         'python-snappy', 'charset-normalizer', 'psutil', 'scipy', 'requests',
-        'networkx','typing-extensions']
+        'networkx','typing-extensions', 'HLL']
```
Comment (JGSweets) [inline .pre-commit-config.yaml:111]: Was this a needed add?

File: dataprofiler/tests/profilers/test_profile_builder.py [hunk 4]
```diff
@@ -1538,6 +1545,63 @@ def test_save_and_load_no_labeler(self):
         save_profile.update_profile(pd.DataFrame({"a": [4, 5]}))
         load_profile.update_profile(pd.DataFrame({"a": [4, 5]}))
 
+    def test_save_json_file(self):
```
Comment (micdavis) [inline dataprofiler/tests/profilers/test_profile_builder.py:1548]: testing save on `StructuredProfiler()` with csv file

File: dataprofiler/profilers/profile_builder.py [hunk 5]
```diff
@@ -1083,7 +1085,7 @@ def _restore_data_labelers(self, data_labeler: BaseDataLabeler = None) -> None:
                 data_labeler_profile = profiler._profiles["data_labeler"]
                 data_labeler_profile.data_labeler = data_labeler
 
-    def _save_helper(self, filepath: str | None, data_dict: dict) -> None:
+    def _pkl_save_helper(self, filepath: str | None, data_dict: dict) -> None:
```
Comment (JGSweets) [inline dataprofiler/profilers/profile_builder.py:1088]: Nice

File: dataprofiler/profilers/profile_builder.py (hunk 1 above)
Comment (taylorfturner) [inline dataprofiler/profilers/profile_builder.py:1549]: wonder if this should be a bit more clear... the `filepath` connotates to me just the folder path... not necessarily the name of the file ... but the filename is vital here with its extension becuause the core logic to testing wether to go with `_pkl_save_helper` or `_json_save_helper` is keying off the extension of the file

File: dataprofiler/profilers/profile_builder.py [hunk 6]
```diff
@@ -1112,6 +1114,19 @@ def _save_helper(self, filepath: str | None, data_dict: dict) -> None:
...
+        """
+        Save profiler to disk.
+
+        :param filepath: Path of file to save to
+        :type filepath: String
+        :return: None
+        """
+        if filepath is None:
+            filepath = ""
+        with open(filepath, "w") as f:
```
Comment (taylorfturner) [inline dataprofiler/profilers/profile_builder.py:1127]: don't we need some kind of file name at least for this because `with open("", "w") as f:`, even if it did succeded wouldn't be ideal for finding the thing it wrote

File: dataprofiler/profilers/profile_builder.py [hunk 7]
```diff
@@ -1112,6 +1114,19 @@ def _save_helper(self, filepath: str | None, data_dict: dict) -> None:
...
+    def _json_save_helper(self, filepath: str) -> None:
+        """
+        Save profiler to disk.
+
+        :param filepath: Path of file to save to
+        :type filepath: String
+        :return: None
+        """
+        if filepath is None:
+            filepath = ""
```
Comment (taylorfturner) [inline dataprofiler/profilers/profile_builder.py:1126]:
```suggestion
            if filepath is None:
              filepath = "profile-{}.json".format(
                  datetime.now().strftime("%d-%b-%Y-%H:%M:%S.%f")
              )
```

File: dataprofiler/profilers/profile_builder.py (hunk 6 above)
Comment (taylorfturner) [reply dataprofiler/profilers/profile_builder.py:1127]: see suggestion above

File: dataprofiler/profilers/profile_builder.py (hunk 1 above)
Comment (taylorfturner) [reply dataprofiler/profilers/profile_builder.py:1549]: we allow for `None`, yes; but the code in `_pkl_save_helper` auto creates a `filepath` (really it is create a filename) automatically with a timestamp for uniquness

File: dataprofiler/tests/profilers/test_profile_builder.py [hunk 8]
```diff
@@ -1538,6 +1545,63 @@ def test_save_and_load_no_labeler(self):
...
+        self.maxDiff = None
```
Comment (taylorfturner) [inline dataprofiler/tests/profilers/test_profile_builder.py:1549]: remove

Comment (taylorfturner) [inline dataprofiler/tests/profilers/test_profile_builder.py:1549]:
```suggestion
```

File: dataprofiler/tests/profilers/test_profile_builder.py [hunk 9]
```diff
@@ -3834,6 +3898,19 @@ def test_save_and_load_no_labeler(self):
...
+
+        # Create Data and StructuredProfiler objects
+        data = dp.Data(os.path.join(data_folder, test_file))
+        save_profile = UnstructuredProfiler(data)
+
+        with self.assertRaisesRegex(
+            NotImplementedError, "UnstructuredProfiler serialization not supported."
+        ):
+            save_profile.save("out.json")
+
```
Comment (taylorfturner) [inline dataprofiler/tests/profilers/test_profile_builder.py:3913]: nice

File: dataprofiler/tests/profilers/test_profile_builder.py [hunk 10]
```diff
@@ -1538,6 +1545,63 @@ def test_save_and_load_no_labeler(self):
...
+        options = ProfilerOptions()
+        options.set({"correlation.is_enabled": True})
+        with test_utils.mock_timeit():
+            save_profile = dp.StructuredProfiler(data)
+
+        # Save and Load profile with Mock IO
+        with mock.patch("builtins.open") as m:
+            mock_file = setup_save_mock_string_open(m)
+            save_profile.save("mock.json")
+            mock_file.seek(0)
```
Comment (taylorfturner) [inline dataprofiler/tests/profilers/test_profile_builder.py:1564]: two thoughts (not hard and fast here), but we should validate the return_value from `json.dumps()` here... I say that because the error in the code above would have been caught. Maybe there is a return_value from `json.dumps()` that we could validate the `filepath` variable that is being actually written... we'd have to mock I reckon

**Review by taylorfturner — CHANGES_REQUESTED** [review]
I think there is a 🐛 (where the test shouldn't pass / we are missing assert / validation) and a couple questions / thoughts for improvement

File: dataprofiler/profilers/profile_builder.py (hunk 7 above)
Comment (taylorfturner) [reply dataprofiler/profilers/profile_builder.py:1126]: this is sneaky because it probably silently runs but doesn't do what we actually want ... 💀

File: dataprofiler/profilers/profile_builder.py (hunk 6 above)
Comment (taylorfturner) [reply dataprofiler/profilers/profile_builder.py:1127]: this is sneaky because it probably silently runs but doesn't do what we actually want ... 💀

File: dataprofiler/tests/profilers/test_profile_builder.py (hunk 10 above)
Comment (taylorfturner) [reply dataprofiler/tests/profilers/test_profile_builder.py:1564]: or just add a test validating passing `None` to the `filepath` parameter... and validate it outputs the autogenerate filepath (would have to mock the time generation)

File: dataprofiler/profilers/profile_builder.py [hunk 11]
```diff
@@ -1112,6 +1114,19 @@ def _save_helper(self, filepath: str | None, data_dict: dict) -> None:
         # Restore all data labelers
         self._restore_data_labelers(data_labelers)
 
+    def _json_save_helper(self, filepath: str) -> None:
```
Comment (JGSweets) [inline dataprofiler/profilers/profile_builder.py:1117]: if we decide to keep this format and we allow `filepath` to be `None`, this typing needs to have a `| None`

File: dataprofiler/profilers/profile_builder.py (hunk 6 above)
Comment (JGSweets) [reply dataprofiler/profilers/profile_builder.py:1127]: good callout

File: dataprofiler/profilers/profile_builder.py (hunk 1 above)
Comment (JGSweets) [reply dataprofiler/profilers/profile_builder.py:1549]: We might want to use python's pathlib as it auto will get for us properly and handle strange cases like:
```
import pathlib


print(pathlib.Path("hello/foo.bar.tar.gz").suffix)
# output: ".gz"

 print(pathlib.Path("hello/foo.bar.tar.gz/a").suffix)
# output: ""
```
notably the last case which has `.` but not as the ext

Comment (JGSweets) [reply dataprofiler/profilers/profile_builder.py:1549]: I think the logic we might is more like:
1. use `type` first to determine how to save
    a. basically can force save with any extension to any type 
2. if `type=None` look to extension to save
3. if extension is not clear, choose pickle
4. if filepath not specified, choose pickle

File: dataprofiler/profilers/profile_builder.py [hunk 12]
```diff
@@ -1531,21 +1546,24 @@ def save(self, filepath: str = None) -> None:
         :type filepath: String
         :return: None
         """
-        # Create dictionary for all metadata, options, and profile
```
Comment (JGSweets) [inline dataprofiler/profilers/profile_builder.py:1534]: above in `save` we need to add the `type: str | None` with default `None` where we can choose '"pickle "| "json"`

File: dataprofiler/profilers/profile_builder.py [hunk 13]
```diff
@@ -1112,6 +1114,19 @@ def _save_helper(self, filepath: str | None, data_dict: dict) -> None:
...
+        :param filepath: Path of file to save to
+        :type filepath: String
+        :return: None
+        """
+        if filepath is None:
+            filepath = ""
+        with open(filepath, "w") as f:
+            json.dump(self, f, cls=ProfileEncoder)
+
     def save(self, filepath: str = None) -> None:
```
Comment (JGSweets) [inline dataprofiler/profilers/profile_builder.py:1130]: Here we need to add the type: str | None with default None where we can choose '"pickle "| "json"`

File: dataprofiler/profilers/profile_builder.py (hunk 7 above)
Comment (JGSweets) [reply dataprofiler/profilers/profile_builder.py:1126]: good call, should be same as pkl technique.

File: dataprofiler/profilers/profile_builder.py (hunk 1 above)
Comment (taylorfturner) [reply dataprofiler/profilers/profile_builder.py:1549]: > use type first to determine how to save

agree

File: dataprofiler/profilers/profile_builder.py [hunk 14]
```diff
@@ -1112,14 +1115,32 @@ def _save_helper(self, filepath: str | None, data_dict: dict) -> None:
...
+    def _json_save_helper(self, filepath: str | None) -> None:
         """
         Save profiler to disk.
 
         :param filepath: Path of file to save to
         :type filepath: String
         :return: None
         """
+        if filepath is None:
+            filepath = "profile-{}.pkl".format(
```
Comment (JGSweets) [inline dataprofiler/profilers/profile_builder.py:1127]: Fix extension here

**Review by JGSweets — CHANGES_REQUESTED** [review]

File: dataprofiler/profilers/profile_builder.py (hunk 14 above)
Comment (taylorfturner) [reply dataprofiler/profilers/profile_builder.py:1127]: good catch

File: dataprofiler/tests/profilers/test_profile_builder.py [hunk 15]
```diff
@@ -1538,6 +1545,113 @@ def test_save_and_load_no_labeler(self):
...
+        mock_open.assert_called_with(expected_second_path, "w")
+
+    @mock.patch(
+        "dataprofiler.profilers.data_labeler_column_profile." "DataLabelerColumn.update"
+    )
+    @mock.patch(
+        "dataprofiler.profilers.profile_builder.DataLabeler",
+        spec=BaseDataLabeler,
+    )
+    def test_save_value_error(self, *mocks):
```
Comment (taylorfturner) [inline dataprofiler/tests/profilers/test_profile_builder.py:1635]:
```suggestion
    def test_save_structured_profiler_value_error(self, *mocks):
```

**Review by taylorfturner — CHANGES_REQUESTED** [review]

File: dataprofiler/tests/profilers/test_profile_builder.py (hunk 15 above)
Comment (micdavis) [reply dataprofiler/tests/profilers/test_profile_builder.py:1644]: there is a testing class that specifies this, resolving

Comment (micdavis) [reply dataprofiler/tests/profilers/test_profile_builder.py:1644]: and all comments below will be resolve for same reason

File: .pre-commit-config.yaml (hunk 3 above)
Comment (taylorfturner) [reply .pre-commit-config.yaml]: removed

File: dataprofiler/profilers/profile_builder.py (hunk 12 above)
Comment (micdavis) [reply dataprofiler/profilers/profile_builder.py:1536]: done

File: dataprofiler/profilers/profile_builder.py [hunk 16]
```diff
@@ -1114,14 +1116,32 @@ def _save_helper(self, filepath: str | None, data_dict: dict) -> None:
...
+        with open(filepath, "w") as f:
+            json.dump(self, f, cls=ProfileEncoder)
+
+    def save(self, filepath: str = None, save_method: str = "pickle") -> None:
+        """
+        Save profiler to disk.
+
+        :param filepath: Path of file to save to
+        :type filepath: String
+        :param save_method: the desired saving method (must be "pickle" or "json")
```
Comment (taylorfturner) [inline dataprofiler/profilers/profile_builder.py:1141]:
```suggestion
        :param save_method: The desired saving method (must be "pickle" or "json")
```

**Review by taylorfturner — APPROVED** [review]

**Review by tyfarnan — APPROVED** [review]

File: dataprofiler/tests/profilers/test_profile_builder.py [hunk 17]
```diff
@@ -1538,6 +1545,130 @@ def test_save_and_load_no_labeler(self):
...
+        )
+
+        # Create Data and StructuredProfiler objects
+        with test_utils.mock_timeit():
+            save_profile = dp.StructuredProfiler(df_structured, options=profile_options)
+
+        # Save and Load profile with Mock IO
+        with mock.patch("builtins.open") as mock_open, mock.patch(
+            "dataprofiler.profilers.profile_builder." "datetime"
+        ) as time_out:
```
Comment (JGSweets) [inline dataprofiler/tests/profilers/test_profile_builder.py:1583]: really should clarify any variable that is a mock as a mock.

e.g. `mock_time_now`

File: dataprofiler/tests/profilers/test_profile_builder.py [hunk 18]
```diff
@@ -1538,6 +1545,130 @@ def test_save_and_load_no_labeler(self):
...
+            }
+        )
+
+        # Create Data and StructuredProfiler objects
+        with test_utils.mock_timeit():
+            save_profile = dp.StructuredProfiler(df_structured, options=profile_options)
+
+        # Save and Load profile with Mock IO
+        with mock.patch("builtins.open") as mock_open, mock.patch(
+            "dataprofiler.profilers.profile_builder." "datetime"
```
Comment (JGSweets) [inline dataprofiler/tests/profilers/test_profile_builder.py:1582]: `"dataprofiler.profilers.profile_builder.datetime"

File: dataprofiler/tests/profilers/test_profile_builder.py [hunk 19]
```diff
@@ -1538,6 +1545,130 @@ def test_save_and_load_no_labeler(self):
...
+            time_out.now().strftime.return_value = "now"
+            setup_save_mock_string_open(mock_open)
+            save_profile.save(save_method="json")
+
+        expected_second_path = "profile-now.json"
+
+        mock_open.assert_called_with(expected_second_path, "w")
+
+    @mock.patch(
+        "dataprofiler.profilers.data_labeler_column_profile." "DataLabelerColumn.update"
```
Comment (JGSweets) [inline dataprofiler/tests/profilers/test_profile_builder.py:1638]: "dataprofiler.profilers.data_labeler_column_profile.DataLabelerColumn.update"

File: dataprofiler/tests/profilers/test_profile_builder.py [hunk 20]
```diff
@@ -1538,6 +1545,130 @@ def test_save_and_load_no_labeler(self):
...
+        }
+
+        actual_data = json.loads(mock_file.read())
+
+        mock_open.assert_called_with(expected_first_path, "w")
+        self.assertDictEqual(expected_data, actual_data)
+
+        # do a second call without a specified file path
+        with mock.patch("builtins.open") as mock_open, mock.patch(
+            "dataprofiler.profilers.profile_builder." "datetime"
```
Comment (JGSweets) [inline dataprofiler/tests/profilers/test_profile_builder.py:1627]: "dataprofiler.profilers.profile_builder.datetime"
